### PR TITLE
added water consumption evaluation 

### PR DESCRIPTION
**Pull request by HsuChieh**

### CODE REVIEW

File: ecologits/data/electricity_mixes.csv [hunk 1]
```diff
@@ -1,5 +1,5 @@
-name,adpe,pe,gwp
-WOR,0.0000000737708,9.988,0.590478
+name,adpe,pe,gwp,wcf
```
Comment (samuelrince) [inline ecologits/data/electricity_mixes.csv:1]: This should be `wue` instead of `wcf`?

File: ecologits/_ecologits.py [hunk 2]
```diff
@@ -131,7 +132,15 @@ def init(
...
         if isinstance(providers, str):
+            EcoLogits.config.provider_selected = providers
             providers = [providers]
+        if EcoLogits.config.provider_selected == "litellm":
+            warnings.warn(
+                "With litellm, the water usage efficiency (WUE) and power usage efficiency (PUE)"
+                "of Amazon Web Services are used",
+                UserWarning,
+                stacklevel=2
+            )
```
Comment (samuelrince) [inline ecologits/_ecologits.py:143]: This should not be necessary for LiteLLM. We use the real provider to compute the impacts (see function `litellm_match_model` in `litellm_tracer.py`)

File: ecologits/_ecologits.py [hunk 3]
```diff
@@ -131,7 +132,15 @@ def init(
             electricity_mix_zone: ISO 3166-1 alpha-3 code of the electricity mix zone (WOR by default).
         """
         if isinstance(providers, str):
+            EcoLogits.config.provider_selected = providers
```
Comment (samuelrince) [inline ecologits/_ecologits.py:135]: This parameter is confusing. What if we initialize with multiple providers?

File: ecologits/impacts/llm.py [hunk 4]
```diff
@@ -2,8 +2,9 @@
 from math import ceil
 from typing import Any, Optional, Union, cast
 
+from ecologits._ecologits import EcoLogits
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:5]: This module should not depend on `_ecologits.py` or we risk creating circular dependencies. This module should use low-level only parameters only. Meaning it is not aware of electricity mixes per location, the list of providers and models, etc. Only the raw formulas.

File: ecologits/impacts/llm.py [hunk 5]
```diff
@@ -28,7 +29,63 @@
...
+    "cohere"	: "AWS",
+    "databricks" : "Microsoft",
+    "meta"	: "Meta",
+    "azureopenai" : "Microsoft", #treated the same way as OpenAI
+    "huggingface_hub" : "AWS",
+    "google" : "Google",
+    "microsoft"	: "Microsoft",
+    "openai" : "Microsoft",
+    "litellm" : "AWS" #need a way to identify provider from model inputed
+}
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:66]: Do we have sources for this? I am not a fan of hard-coding cloud provider names only to retrieve a PUE and a WUE. I'd rather have one value by AI provider (OpenAI, Mistral AI, ...) and explain in the docs that we use the PUE/WUE of Microsoft for OpenAI or an EU region average for Mistral AI, etc...

File: ecologits/impacts/llm.py [hunk 6]
```diff
@@ -150,7 +207,9 @@ def server_energy(
 
 @dag.asset
 def request_energy(
-        datacenter_pue: float,
+        provider: str,
+        provider_pue: dict,
+        ai_company_to_data_center_provider: dict,
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:212]: As mentioned above, keep this simple; only use the `datacenter_pue` as input. This value of PUE will be retrieved before this function call with the right provider.

File: ecologits/impacts/llm.py [hunk 7]
```diff
@@ -350,14 +444,44 @@ def request_embodied_pe(
...
 def compute_llm_impacts_dag(
+        provider: str,
         model_active_parameter_count: ValueOrRange,
         model_total_parameter_count: ValueOrRange,
         output_token_count: float,
         request_latency: float,
         if_electricity_mix_adpe: float,
         if_electricity_mix_pe: float,
         if_electricity_mix_gwp: float,
+        if_electricity_mix_wcf: float,
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:484]: Should be `if_electricity_mix_wue`?

File: ecologits/impacts/llm.py [hunk 8]
```diff
@@ -28,7 +29,63 @@
...
+    "meta"	: "Meta",
+    "azureopenai" : "Microsoft", #treated the same way as OpenAI
+    "huggingface_hub" : "AWS",
+    "google" : "Google",
+    "microsoft"	: "Microsoft",
+    "openai" : "Microsoft",
+    "litellm" : "AWS" #need a way to identify provider from model inputed
+}
+
+BATCHING_SIZE = 16
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:68]: As we are working on integrating the batch_size, I would do this in another PR.

File: ecologits/impacts/modeling.py [hunk 9]
```diff
@@ -130,6 +130,23 @@ class PE(BaseImpact):
     name: str = "Primary Energy"
     unit: str = "MJ"
 
+class Water(BaseImpact):
```
Comment (samuelrince) [inline ecologits/impacts/modeling.py:133]: Should we `WCF` instead? If we use GWP, ADPe, PE for other impacts which are technical names, I'd rather keep that convention. This does not mean we should simplify this for the end-user, but it is another PR.

File: docs/index.md [hunk 10]
```diff
@@ -192,6 +192,7 @@ And, **multiple criteria**:
 * **Global Warming Potential** (GWP): related to climate change, commonly known as GHG emissions in kgCO2eq.
 * **Abiotic Depletion Potential for Elements** (ADPe): related to the depletion of minerals and metals in kgSbeq.
 * **Primary Energy** (PE): related to the energy consumed from primary sources like oil, gas or coal in MJ. 
+* **Water Use** (water): Water consumption from this request.
```
Comment (samuelrince) [inline docs/index.md:195]: What about "Water Consumption Footprint" (WCF) to stay coherent with the other impacts?

File: .gitignore [hunk 11]
```diff
@@ -162,4 +162,9 @@ dmypy.json
...
 
 # Precommit hooks: ruff cache
-.ruff_cache
\ No newline at end of file
+.ruff_cache
+
+#Chieh encountered issues when using pre-commit-config. 
+#The modified files should not be committed. 
+run_safety.bat
+.pre-commit-config.yaml.pre-commit-config.yaml
```
Comment (samuelrince) [inline .gitignore:170]: What are these files exactly? Can you just not add them to tracked files?

**Review by samuelrince — CHANGES_REQUESTED** [review]

File: ecologits/data/electricity_mixes.csv (hunk 1 above)
Comment (HsuChieh) [reply ecologits/data/electricity_mixes.csv:1]: In the World Resources Intitute report, it's marked as WCF so I used it as is, but fonctionally it's the same thing as WUE off-site. What do you suggest I do about it?

File: ecologits/_ecologits.py (hunk 2 above)
Comment (HsuChieh) [reply ecologits/_ecologits.py:143]: It's something I haven't figured out... I created both these parameters, and I still need to find a way to impliment this for litellm

File: ecologits/_ecologits.py (hunk 3 above)
Comment (HsuChieh) [reply ecologits/_ecologits.py:135]: I can try.

File: ecologits/impacts/llm.py (hunk 5 above)
Comment (HsuChieh) [reply ecologits/impacts/llm.py:66]: I'll make it more robust.

**Comment by HsuChieh** [comment]
@samuelrince should be fine. I'm more busy in the month of August, but I'll try to resolve these issues in the next two weeks or so.

File: ecologits/impacts/modeling.py (hunk 9 above)
Comment (HsuChieh) [reply ecologits/impacts/modeling.py:133]: Done

File: docs/index.md (hunk 10 above)
Comment (HsuChieh) [reply docs/index.md:195]: done

File: ecologits/impacts/llm.py (hunk 5 above)
Comment (HsuChieh) [reply ecologits/impacts/llm.py:66]: in documentation, check all documention

**Comment by samuelrince** [comment]
Hey @HsuChieh, 

I have a few more comments before merging! 

I cannot ignore commits while merging the PR on GitHub, it will take the state of your entire branch. Some recommendations to make this easier (you can find other information [here](https://ecologits.ai/latest/contributing/#pull-requests) as well).

I will need you to make some changes on your fork in order to properly merge this PR first. The objective is to have a clean history with the work you have implemented that does not conflict with the current state of the project. I give a plan of action below; tell me if something is unclear. I'd be happy to jump in a meeting to help you!


**Plan:**

1. Move your work on branch different from main (e.g. `feat/water-consumption`)
2. Synchronize your `main` branch with the source (this will remove your commits from the main branch)
3. Rebase your `feat/water-consumption` branch onto your now synchronized `main`. You'll probably need to fix some conflicts.
4. Push force the changes on your remote branch `feat/water-consumption`, this update this PR and make the merge easy.
5. Drop commits that are not relevant to this PR. (probably fb06558, c4fbe15, 1efc4fc and 29e3598)

Also, I've changed the target branch to `dev/methodology-update` because this will be in the release 0.9.0, I don't want it to interfere with patches to the 0.8.0. At the end you should have zero commits behind, only commits ahead of the target branch.

PS: If you have conflicts with the `poetry.lock` file, do not try to fix it by hand. ALWAYS accept the source version first, then run a `poetry lock` command to update, and finally push the updated `poetry.lock` file.

File: docs/index.md [hunk 12]
```diff
@@ -198,6 +198,7 @@ And, **multiple criteria**:
 * **Global Warming Potential** (GWP): related to climate change, commonly known as GHG emissions in kgCO2eq.
 * **Abiotic Depletion Potential for Elements** (ADPe): related to the depletion of minerals and metals in kgSbeq.
 * **Primary Energy** (PE): related to the energy consumed from primary sources like oil, gas or coal in MJ. 
+* **Water Use** (WCF): Water consumption from this request.
```
Comment (samuelrince) [inline docs/index.md:201]: Should be "Water Consumption Footprint".

File: docs/methodology/index.md [hunk 13]
```diff
@@ -16,11 +16,12 @@ Upcoming methodologies ([join us](https://genai-impact.org/contact) to help spee
 
 EcoLogits employs the **Life Cycle Assessment (LCA) methodology**, as defined by ISO 14044, to estimate the environmental impacts of requests made to generative AI inference services. This approach focuses on **multiple phases** of the lifecycle, specifically raw material extraction, manufacturing, transportation (denoted as embodied impacts), usage and end-of-life. Notably, we do not cover the end-of-life phase due to data limitations on e-waste recycling.
 
-Our assessment considers **three key environmental criteria**:
+Our assessment considers **four key environmental criteria**:
 
 - Global Warming Potential (GWP): Evaluates the impact on global warming in terms of CO2 equivalents.
 - Abiotic Resource Depletion for Elements (ADPe): Assesses the consumption of raw minerals and metals, expressed in antimony equivalents.
 - Primary Energy (PE): Calculates energy consumed from natural sources, expressed in megajoules.
+- Water Use (WCF): Water consumption from this request.
```
Comment (samuelrince) [inline docs/methodology/index.md:24]: "Water Consumption Footprint"

File: docs/methodology/llm_inference.md [hunk 14]
```diff
@@ -344,7 +417,12 @@ We aim at covering the largest scope possible when assessing the environmental i
 Please cite **GenAI Impact** non-profit organization and **link to this documentation page**. 
 
 ```bibtex
-Coming soon...
+"""
+@software{ecologits,
+  author = {Samuel Rincé, Adrien Banse, Vinh Nguyen, Luc Berton, and Chieh Hsu},
+  publisher = {GenAI Impact},
+  title = {EcoLogits: track the energy consumption and environmental footprint of using generative AI models through APIs.},
+}"""
```
Comment (samuelrince) [inline docs/methodology/llm_inference.md:425]: The citation block was for the paper about the methodology, but we never wrote it (at least for now). I think we can simply point to the GitHub repository to get the citation. I think @adrienbanse set this up.

File: ecologits/impacts/llm.py [hunk 15]
```diff
@@ -28,7 +28,55 @@
...
+    "cohere"	: 0.18, #2023
+    "databricks" : 0.49, #2022
+    "meta"	: 0.18,    # L/kWh, 2023
+    "azureopenai" : 0.49, #2022 #treated the same way as OpenAI
+    "huggingface_hub" : 0.18, #2023
+    "google" : 0.916,
+    "microsoft"	: 0.49, #2022
+    "openai" : 0.49, #2022
+    "litellm" : 0.18, #2023 #need a way to identify provider from model inputed
+}
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:57]: This should preferably be in the `tracers/utils.py` module. Match the provider with these mappings to then input the correct `datacenter_pue` and `datacenter_wue` values in the `compute_llm_impacts` function.

File: ecologits/impacts/llm.py [hunk 16]
```diff
@@ -28,7 +28,55 @@
...
+
+BATCHING_SIZE = 1 #Using 1 as placeholder. Will be changed once more research on batching size is done.
+
+GPUS_IN_SERVER = 8
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:61]: We have a constant that already counts the number of GPUs: `SERVER_GPUS`.

File: ecologits/impacts/llm.py [hunk 17]
```diff
@@ -28,7 +28,55 @@
...
+# 15mm x 15mm chip: 225 mm²
+# So, the simple calculation would be:
+
+# 70,685 mm2 / 225 mm2 ​≈ 314 chips
+
+# https://esg.tsmc.com/en-US/file/public/e-all_2023.pdf
+# page 114, 2023
+# 2023 - 176.4 Water consumption per wafer-layer (Liter/12-inch equivalent wafer mask layer)
+# 176.4/314 =
+# 0.56178343949 L/chip
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:78]: If these details are already in the docs you can remove the comments.

File: ecologits/impacts/llm.py [hunk 18]
```diff
@@ -150,7 +198,8 @@ def server_energy(
 
 @dag.asset
 def request_energy(
-        datacenter_pue: float,
+        provider: str,
+        ai_provider_pue: dict,
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:202]: Remove `provider` and use `datacenter_pue` instead.

File: ecologits/impacts/llm.py [hunk 19]
```diff
@@ -223,6 +275,35 @@ def request_usage_pe(
     """
     return request_energy * if_electricity_mix_pe
 
+@dag.asset
+def request_usage_water(
+        request_energy: ValueOrRange,
+        if_electricity_mix_wue: float,
+        ai_provider_wue_on_site: dict,
+        provider: str,
+        ai_provider_pue: dict
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:284]: Remove `provider` and rename to `datacenter_pue` and `datacenter_wue`

File: ecologits/impacts/llm.py [hunk 20]
```diff
@@ -350,14 +431,44 @@ def request_embodied_pe(
     return (generation_latency / server_lifetime) * server_gpu_embodied_pe
 
 
+@dag.asset
+def request_embodied_water(
+        server_lifetime: float,
+        batching_size: float,
+        water_fabricating_gpu: float,
+        gpus_in_server: float,
+        generation_latency: ValueOrRange
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:440]: - Reuse `server_gpu_count` instead of `gpus_in_server`
- `water_fabricating_gpu` should be `gpu_embodied_wcf` (look at the other impacts).

File: ecologits/impacts/llm.py [hunk 21]
```diff
@@ -350,14 +431,44 @@ def request_embodied_pe(
...
+    """
+
+    output = generation_latency *water_fabricating_gpu * gpus_in_server/ (server_lifetime * batching_size)
+
+    return output
+
+
+
 def compute_llm_impacts_dag(
+        provider: str,
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:463]: Remove `provider`

File: ecologits/impacts/llm.py [hunk 22]
```diff
@@ -375,19 +486,25 @@ def compute_llm_impacts_dag(
         server_embodied_adpe: Optional[float] = SERVER_EMBODIED_IMPACT_ADPE,
         server_embodied_pe: Optional[float] = SERVER_EMBODIED_IMPACT_PE,
         server_lifetime: Optional[float] = HARDWARE_LIFESPAN,
-        datacenter_pue: Optional[float] = DATACENTER_PUE,
+        ai_provider_wue_on_site: Optional[dict] = AI_PROVIDER_WUE_ON_SITE,
+        ai_provider_pue: Optional[dict] = AI_PROVIDER_PUE,
+        water_fabricating_gpu: Optional[float] = WATER_FABRICATING_GPU,
+        gpus_in_server: Optional[float] = GPUS_IN_SERVER,
+        batching_size: Optional[float] =  BATCHING_SIZE
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:493]: Use `datacenter_pue` and `datacenter_pue` instead and remove `gpus_in_server`.

File: ecologits/impacts/llm.py [hunk 23]
```diff
@@ -435,31 +557,38 @@ def compute_llm_impacts_dag(
...
+        ai_provider_pue=ai_provider_pue,
+        water_fabricating_gpu=water_fabricating_gpu,
+        gpus_in_server=gpus_in_server,
+        batching_size=batching_size
     )
     return results
 
-
 def compute_llm_impacts(
+        provider: str,
```
Comment (samuelrince) [inline ecologits/impacts/llm.py:569]: No need for `provider` there as well.

**Review by samuelrince — CHANGES_REQUESTED** [review]

File: ecologits/impacts/llm.py (hunk 20 above)
Comment (HsuChieh) [reply ecologits/impacts/llm.py]: that's not it. it's the amount of water required to fabricate a gpu, different from gpu_embodied_wcf
